src/main/java/com/google/api/gax/bundling/ThresholdBundler.java

```diff
@@ -117,204 +135,106 @@ private Builder() {
     }
   }
 
-  /** Get a new builder for a ThresholdBundler. */
-  public static <T> Builder<T> newBuilder() {
-    return new Builder<T>();
+  public static <E> Builder<E> newBuilder() {
+    return new Builder<>();
   }
 
   /**
    * Adds an element to the bundler. If the element causes the collection to go past any of the
-   * thresholds, the bundle will be made available to consumers.
-   *
-   * @throws FlowControlException
+   * thresholds, the bundle will be sent to the {@code ThresholdBundleReceiver}.
    */
   public void add(E e) throws FlowControlException {
-    final Lock lock = this.lock;
     // We need to reserve resources from flowController outside the lock, so that they can be
-    // released by removeBundle().
+    // released by pushCurrentBundle().
     flowController.reserve(e);
     lock.lock();
     try {
-      boolean signalBundleIsReady = false;
-      if (currentBundleState == null) {
-        currentBundleState = new BundleState(thresholdPrototypes, maxDelay);
-        currentBundleState.start();
-        signalBundleIsReady = true;
-      }
-
-      currentBundleState.add(e);
-      if (currentBundleState.isAnyThresholdReached()) {
-        signalBundleIsReady = true;
-        closedBundles.add(currentBundleState.getBundle());
-        currentBundleState = null;
+      receiver.validateBundle(e);
+      boolean anyThresholdReached = isAnyThresholdReached(e);
+
+      if (currentOpenBundle == null) {
+        currentOpenBundle = e;
+        // Schedule a job only when no thresholds have been exceeded, otherwise it will be
+        // immediately cancelled
+        if (!anyThresholdReached) {
+          currentAlarmFuture =
+              executor.schedule(
+                  pushCurrentBundleRunnable, maxDelay.getMillis(), TimeUnit.MILLISECONDS);
+        }
+      } else {
+        bundleMerger.merge(currentOpenBundle, e);
       }
 
-      if (signalBundleIsReady) {
-        bundleCondition.signalAll();
+      if (anyThresholdReached) {
+        pushCurrentBundle();
       }
     } finally {
       lock.unlock();
     }
   }
 
   /**
-   * Makes the currently contained elements available for consumption, even if no thresholds were
-   * triggered.
+   * * Package-private for use in testing.
    */
-  public void flush() {
-    final Lock lock = this.lock;
+  boolean isEmpty() {
     lock.lock();
     try {
-      if (currentBundleState != null) {
-        closedBundles.add(currentBundleState.getBundle());
-        currentBundleState = null;
-      }
-      bundleCondition.signalAll();
+      return currentOpenBundle == null;
     } finally {
       lock.unlock();
     }
   }
 
   /**
-   * Remove and return the current bundle, regardless of whether it has triggered any thresholds.
+   * Push the current bundle to the bundle receiver. Returns an ApiFuture<Void> that completes once
+   * the bundle has been processed by the bundle receiver and the flow controller resources have
+   * been released.
```

Comment:
Done.

As discussed with @pongad, we will not support flushing in the bundling implementation at this time. If in the future we decide to (a) make the bundling surface visible to users, and (b) add flushing, we would implement it by holding a reference to all outstanding ApiFutures, and allowing flush to wait until all are complete.

For APIs where flushing is required, such as logging, we will implement flushing on top of bundling using the individual log entry futures.